### PR TITLE
Add dataServicesCall mediator

### PR DESCRIPTION
**Pull request by chanikag**
This PR is to add dataServicesCall mediator to directly call dataservices from the synapse mediation layer.

This is introduced to overcome following issues currently we have when using call mediator after creating a payload using payload factory mediator and setting up necessary headers (ex: SOAPAction header), even though the dataservice features and mediation features are shipped under same EI profile.

1. Invoking through HTTP transport adds network latency and that impacts on the performance.
2. The payload has to be created separately and that creates an overhead in the deployment time.

Here is the syntax of this mediator,

For single operations:
```
      <dataServicesCall serviceName="RDBMSSample">
            <operations type="single_req">
               <operation name="addEmployee">
                  <param name="employeeNumber" type="int" value="2"></param>
                  <param name="firstname" type="string" value="Peter"></param>
                  <param name="lastName" type="string" value="Parker"></param>
                  <param name="email" type="string" value="peter@wso2.com"></param>
                  <param name="salary" type="int" value="12"></param>
               </operation>
            </operations>
            <target type="envelope"/>
         </dataServicesCall>
```

For batch requests:
```
<dataServicesCall serviceName="RDBMSSample">
            <operations type="batch_req">
               <operation name="addEmployee">
                  <param name="employeeNumber" type="int" value="555"></param>
                  <param name="firstname" type="string" value="Peter"></param>
                  <param name="lastName" type="string" value="Parker"></param>
                  <param name="email" type="string" value="peter@wso2.com"></param>
                  <param name="salary" type="int" value="12"></param>
               </operation>
               <operation name="addEmployee">
                  <param name="employeeNumber" type="int" value="444"></param>
                  <param name="firstname" type="string" value="Peter"></param>
                  <param name="lastName" type="string" value="Parker"></param>
                  <param name="email" type="string" value="peter@wso2.com"></param>
                  <param name="salary" type="int" value="12"></param>
               </operation>
            </operations>
            <target type="envelope"/>
         </dataServicesCall>
```
  
For request_box operations:
```
       <dataServicesCall serviceName="RDBMSSample">
            <operations type="request_box">
               <operation name="addEmployee">
                  <param name="employeeNumber" type="string" type="int" value="555"></param>
                  <param name="firstname" type="string" value="Peter"></param>
                  <param name="lastName" type="string" value="Parker"></param>
                  <param name="email" type="string" value="peter@wso2.com"></param>
                  <param name="salary" type="int" value="12"></param>
               </operation>
               <operation name="employeesByNumber">
                  <param name="employeeNumber" type="int" value="555"></param>
               </operation>
            </operations>
            <target type="envelope"/>
         </dataServicesCall>

```

### CODE REVIEW

File: components/mediators/dataservice/org.wso2.carbon.mediator.dataservice/hs_err_pid32185.log [hunk 1]
```diff
@@ -0,0 +1,1012 @@
...
+7f4adf813000-7f4adf81e000 r--s 000c0000 08:05 3674530                    /home/chanika/.m2/repository/org/wso2/carbon/org.wso2.carbon.registry.core/4.4.40/org.wso2.carbon.registry.core-4.4.40.jar
+7f4adf81e000-7f4adf821000 r--s 00015000 08:05 3674524                    /home/chanika/.m2/repository/org/wso2/carbon/javax.cache.wso2/4.4.40/javax.cache.wso2-4.4.40.jar
+7f4adf821000-7f4adf879000 r--s 003a7000 08:05 3674523                    /home/chanika/.m2/repository/org/bouncycastle/bcprov-jdk15on/1.60/bcprov-jdk15on-1.60.jar
+7f4adf879000-7f4adf87b000 r--s 0036d000 08:05 3674522                    /home/chanika/.m2/repository/org/wso2/orbit/org/bouncycastle/bcprov-jdk15on/1.60.0.wso2v1/bcprov-jdk15on-1.60.0.wso2v1.jar
+7f4adf87b000-7f4adf87c000 r--s 00005000 08:05 3674520                    /home/chanika/.m2/repository/org/eclipse/equinox/org.eclipse.equinox.http.helper/1.1.0.wso2v1/org.eclipse.equinox.http.helper-1.1.0.wso2v1.jar
+7f4adf87c000-7f4adf880000 r--s 00011000 08:05 3674519                    /home/chanika/.m2/repository/org/eclipse/osgi/org.eclipse.osgi.services/3.3.100.v20130513-1956/org.eclipse.osgi.services-3.3.100.v20130513-1956.jar
```
Comment (Nirothipan) [inline components/mediators/dataservice/org.wso2.carbon.mediator.dataservice/hs_err_pid32185.log]: Shall we remove this file

Comment (chanikag) [reply components/mediators/dataservice/org.wso2.carbon.mediator.dataservice/hs_err_pid32185.log]: Done.

File: components/mediators/dataservice/org.wso2.carbon.mediator.dataservice/pom.xml [hunk 2]
```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0"?>
+<!--
+ ~ Copyright (c) 2005-2010, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (sameeragunarathne) [inline components/mediators/dataservice/org.wso2.carbon.mediator.dataservice/pom.xml]: license year needs to be changed

File: components/mediators/dataservice/org.wso2.carbon.mediator.dataservice/src/main/java/org/wso2/carbon/mediator/dataservice/DataServiceCallMediator.java [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+
+    public boolean mediate(MessageContext messageContext) {
+
+        SynapseLog synLog = getLog(messageContext);
+        String serviceName = getDsName();
+        String targetTypeName = getTargetType();
+        String propertyName = getPropertyName();
+
+        if (synLog.isTraceOrDebugEnabled()) {
+            synLog.traceOrDebug("The invoking data service name : " + getDsName());
```
Comment (Nirothipan) [inline components/mediators/dataservice/org.wso2.carbon.mediator.dataservice/src/main/java/org/wso2/carbon/mediator/dataservice/DataServiceCallMediator.java]:
```suggestion
            synLog.traceOrDebug("Invoking the data service : " + getDsName());
```

**Comment by sameeragunarathne** [comment]
Closing until the tooling support is available.

**Comment by chanikag** [comment]
Closing as this is added to MI

**Pull request closed without merging**